expression/builtin_compare.go

```diff
@@ -1699,6 +1727,31 @@ func (c *compareFunctionClass) refineArgs(ctx sessionctx.Context, args []Express
 	return c.refineArgsByUnsignedFlag(ctx, []Expression{finalArg0, finalArg1})
 }
 
+// see https://github.com/pingcap/tidb/issues/38361 for more details
+func (c *compareFunctionClass) refineNumericConstantCmpDatetime(ctx sessionctx.Context, args []Expression, constArg *Constant, constArgIdx int) []Expression {
+	dt, err := constArg.Eval(chunk.Row{})
+	if err != nil {
+		return args
+	}
+	sc := ctx.GetSessionVars().StmtCtx
+	var timestampDatum types.Datum
+	targetFieldType := types.NewFieldType(mysql.TypeTimestamp)
+	timestampDatum, err = dt.ConvertTo(sc, targetFieldType)
+	if err != nil {
+		return args
+	}
```

Comment:
maybe also need to return original args if `timestampDatum` is null but `dt` is not null

Reply:
Make sense, add it.